core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java

```diff
@@ -602,18 +603,62 @@ private static AnnotatedValueResolver ofQueryParamMap(String name,
                                                           AnnotatedElement annotatedElement,
                                                           AnnotatedElement typeElement, Class<?> type,
                                                           DescriptionInfo description) {
+        final Type valueType = ((ParameterizedType) ((Parameter) typeElement).getParameterizedType())
+                .getActualTypeArguments()[1];
+        final Class<?> rawValueType = ClassUtil.typeToClass(valueType);
+        assert rawValueType != null;
+
+        if (valueType instanceof ParameterizedType && !(rawValueType == Iterable.class ||
+                                                        rawValueType == List.class ||
+                                                        rawValueType == Collection.class ||
+                                                        rawValueType == Set.class)) {
+            throw new IllegalArgumentException(
+                    "Invalid Map value type: " + rawValueType +
+                    " (expected Iterable, List, Collection or Set)");
+        }
+
+        final BiFunction<AnnotatedValueResolver, ResolverContext, Object> biFunction;
+
+        if (Set.class.isAssignableFrom(rawValueType)) {
+            biFunction = (resolver, ctx) -> ctx.queryParams().stream()
+                                               .collect(toImmutableMap(
+                                                       Entry::getKey,
+                                                       e -> ImmutableSet.of(e.getValue()),
+                                                       (existing, replacement) ->
+                                                               ImmutableSet.<String>builder()
```

Comment:
note: This can be probably be improved to not create a set for each element later on but for now I agree simplicity is better as there probably aren't a lot of query parameters.